pufferlib/ocean/artillery3d/artillery3d.h

```diff
@@ -0,0 +1,687 @@
+        Y0 + barrel_length * barrel_direction.y,
+        Z0 + barrel_length * barrel_direction.z
+    };
+
+    DrawCylinderEx(cannon_pos, barrel_end, barrel_radius, barrel_radius, 12, DARKGRAY);
+
+    DrawSphere(cannon_pos, 12.0f, GRAY);
+
+    float vx0 = MUZZLEV * barrel_direction.x;
+    float vy0 = MUZZLEV * barrel_direction.y;
+    float vz0 = MUZZLEV * barrel_direction.z;
+
+    Vector3 prev_point = cannon_pos;
+    int segment_count = 0;
+
+    float x;
+    float y;
+    float z;
+
+    for (float t = TIMESTEP; t < FUSEMULT; t += TIMESTEP) {
+        x = X0 + vx0 * t;
+        y = Y0 + vy0 * t;
+        z = Z0 + vz0 * t - 0.5f * G * t * t;
+
+        if (z < 0 || x < 0 || x > XSIZE || y < 0 || y > YSIZE) break;
+
+        Vector3 current_point = {x, y, z};
+
+        Color trajectory_color = (segment_count % 2 == 0) ? WHITE : LIGHTGRAY;
+        DrawLine3D(prev_point, current_point, trajectory_color);
+
+        prev_point = current_point;
+        segment_count++;
+    }
+
+    if (env->projectile_active) {
+        DrawSphere((Vector3){env->px, env->py, env->pz}, 4.0f, BLACK);
+        if (env->t > env->fire_t + env->fuse_t * FUSEMULT) {
+            DrawSphere((Vector3){env->boom_x, env->boom_y, env->boom_z}, 20.0f, (Color){128, 128, 128, 128});
+            if (env->dist < EXPLRAD) {
+                DrawSphere((Vector3){env->boom_x, env->boom_y, env->boom_z}, 10.0f+20.0f*(env->t-env->fire_t), (Color){255, 0, 0, 128});
+            }
+        }
+    }
+
+    for (int i = 0; i < NUMTARGETS; i++) {
+        if (env->tx[i] > 1.0f) {
+            DrawSphere((Vector3){env->tx[i], env->ty[i], env->tz[i]}, 15.0f, BLUE);
+        }
+    }
+
+    EndMode3D();
+
+    DrawText("Left click + drag: Rotate camera", 10, 10, 16, WHITE);
+    DrawText("Mouse wheel: Zoom in/out", 10, 30, 16, WHITE);
+    DrawText("Tab: Toggle fullscreen", 10, 50, 16, WHITE);
+    DrawText(TextFormat("Score: %.3f", env->score), 10, 70, 20, BLACK);
+
+    EndDrawing();
+}
+
+void init(Artillery3D* env) {
+    env->runs = 0;
+    env->tick = 0;
+    env->t = 0.0f;
+    env->fire_t = 99999.0f;
+    if (env->same_runs < 1) env->same_runs = 1;
+    env->projectile_active = 0;
+    env->projectile_time = 0.0f;
+    env->dist = XSIZE;
+
+    env->inv_x_size = 1.0f / XSIZE;
+    env->inv_y_size = 1.0f / YSIZE;
+    env->inv_z_size = 1.0f / ZSIZE;
+    env->inv_max_shots = 1.0f / MAXSHOTS;
+
+    env->tx = (float*)calloc(NUMTARGETS, sizeof(float));
+    env->ty = (float*)calloc(NUMTARGETS, sizeof(float));
+    env->tz = (float*)calloc(NUMTARGETS, sizeof(float));
+    env->target_vx = (float*)calloc(NUMTARGETS, sizeof(float));
+    env->target_vy = (float*)calloc(NUMTARGETS, sizeof(float));
+    env->target_vz = (float*)calloc(NUMTARGETS, sizeof(float));
+    env->time_target_vanish = (float*)calloc(NUMTARGETS, sizeof(float));
+
+    srand(env->rng + env->i);
+
+    get_random_start(env);
+}
+
+void allocate(Artillery3D* env) {
```

Comment:
Latest template envs etc have this in the .c, since that's the only place you need it. Can just inline obs reward etc. allocs there